bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java

```diff
@@ -0,0 +1,492 @@
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "SunSpec item should either have the address and length configuration set or should been created by auto discovery");
+            return;
+        }
+    }
+
+    /**
+     * Load configuration from main configuration
+     */
+    private Optional<ModelBlock> getAddressFromConfig() {
+        if (!config.isPresent()) {
+            return Optional.empty();
+        }
+        ModelBlock block = new ModelBlock();
+        block.address = config.get().getAddress();
+        block.length = config.get().getLength();
+        return Optional.of(block);
+    }
+
+    /**
+     * Publish the unique address property if it has not been set before
+     */
+    private void publishUniqueAddress(ModelBlock block) {
+        Map<String, String> properties = getThing().getProperties();
+        if (properties.containsKey(PROPERTY_UNIQUE_ADDRESS) && !properties.get(PROPERTY_UNIQUE_ADDRESS).isEmpty()) {
+            logger.debug("Current unique address is: {}", properties.get(PROPERTY_UNIQUE_ADDRESS));
+            return;
+        }
+
+        ModbusEndpointThingHandler handler = getEndpointThingHandler();
+        if (handler == null) {
+            return;
+        }
+        getThing().setProperty(PROPERTY_UNIQUE_ADDRESS, handler.getUID().getAsString() + ":" + block.address);
+    }
+
+    /**
+     * Dispose the binding correctly
+     */
+    @Override
+    public void dispose() {
+        tearDown();
+    }
+
+    /**
+     * Unregister the poll task and release the endpoint reference
+     */
+    private void tearDown() {
+        unregisterPollTask();
+        unregisterEndpoint();
+    }
+
+    /**
+     * Returns the current slave id from the bridge
+     */
+    public int getSlaveId() {
+        return slaveId;
+    }
+
+    /**
+     * Get the endpoint handler from the bridge this handler is connected to
+     * Checks that we're connected to the right type of bridge
+     *
+     * @return the endpoint handler or null if the bridge does not exist
+     */
+    private @Nullable ModbusEndpointThingHandler getEndpointThingHandler() {
+        Bridge bridge = getBridge();
+        if (bridge == null) {
+            logger.debug("Bridge is null");
+            return null;
+        }
+        if (bridge.getStatus() != ThingStatus.ONLINE) {
+            logger.debug("Bridge is not online");
+            return null;
+        }
+
+        ThingHandler handler = bridge.getHandler();
+        if (handler == null) {
+            logger.debug("Bridge handler is null");
+            return null;
+        }
+
+        if (handler instanceof ModbusEndpointThingHandler) {
+            ModbusEndpointThingHandler slaveEndpoint = (ModbusEndpointThingHandler) handler;
+            return slaveEndpoint;
+        } else {
+            logger.debug("Unexpected bridge handler: {}", handler);
+            throw new IllegalStateException();
```

Comment:
Why would this throw an exception rather than return null?

Reply:
Thanks you are now. Changed it for a more coherent behavior.